Implement efficient cluster state snapshots in ECSStateClient

wjbuys opened the pull request:
This uses an `ECSAsyncClient` to implement the `ECSState` interface, and updates the `SingleTaskScheduler` to actually only run a single task.

This also changes the types in the snapshot to be lists of instances/tasks, since the maps weren't very useful in practice.

The integration test is very basic, but we may be able to build on it for further integration tests that require a test cluster.

Tracking issue: #261

#### Licensing

This contribution is under the terms of the Apache 2.0 License: Yes

wjbuys commented:
Looks like there's another ECS client getting started in the Spring tests, modified the build logic to show more failure details in Travis to see what's going on.

wjbuys commented on scheduling-manager/src/main/java/com/amazonaws/blox/scheduling/state/FakeECSStateClient.java:
```diff
@@ -0,0 +1,32 @@
+ * Temporary fake implementation of ECS State client that just takes time and returns an empty
+ * cluster.
+ */
+public class FakeECSStateClient implements ECSState {
```
That's weird, I thought I deleted this file. I'll remove it.

narehayrapetyan commented on scheduling-manager/src/main/java/com/amazonaws/blox/scheduling/manager/ManagerApplication.java:
```diff
@@ -14,33 +14,34 @@
+  "com.amazonaws.blox.scheduling.manager",
+  "com.amazonaws.blox.scheduling.state",
+})
 public class ManagerApplication extends SchedulingApplication {
```
would it be possible to put all the configuration in one package so it's easy to find?

wjbuys replied:
In this case, it's not the configuration that's in a different package, it's the `@Component` annotation on the state stuff. I'm open to any suggestions on how to deal with that.

narehayrapetyan replied:
What do you mean? I think the schedulingapplication configuration is not in com...manager

wjbuys replied:
Oh, wait, dammit github highlighting. Do you mean "why is SchedulingApplication in a different package"? If so, the reason is that it's the shared config between all components of the scheduling-manager package. If I were to put it in any of the subpackages, then it would no longer be clear that it's common to all of them.

An alternative could be to add a separate top-level `com.amazonaws.blox.scheduling.config` package, and add all the config in there. Is that what you were alluding to?

narehayrapetyan commented on scheduling-manager/src/main/java/com/amazonaws/blox/scheduling/scheduler/engine/SingleTaskScheduler.java:
```diff
@@ -19,22 +19,47 @@
 public class SingleTaskScheduler implements Scheduler {
 
   @Override
   public List<SchedulingAction> schedule(ClusterSnapshot snapshot, EnvironmentVersion environment) {
```
you'll have to create a new internal object for environmentversion or something similar because the returned type is now describeEnvironment which contains separate fields. can be done in the refactoring cr

wjbuys replied:
That's what I did for #278

narehayrapetyan commented on scheduling-manager/src/main/java/com/amazonaws/blox/scheduling/scheduler/engine/SingleTaskScheduler.java:
```diff
@@ -19,22 +19,47 @@
+        snapshot
+            .getInstances()
+            .stream()
+            .sorted(Comparator.comparing(ContainerInstance::getArn))
```
do we need to sort? I suppose this is gonna be removed anyway

wjbuys replied:
yes. Just so we don't start the task multiple times on different instances just because the ordering is different. (I had that before, and ended up with a cluster full of random tasks)

narehayrapetyan commented on scheduling-manager/src/main/java/com/amazonaws/blox/scheduling/scheduler/engine/StartTask.java:
```diff
@@ -15,29 +15,37 @@
                 .containerInstances(containerInstanceArn)
                 .taskDefinition(taskDefinitionArn)
+                .group(group)
+                .startedBy("blox")
```
constant

wjbuys replied:
It's a single usage of an unimportant string in a temporary class. I'll extract it, but I'm not convinced it's helpful.

narehayrapetyan commented on scheduling-manager/src/main/java/com/amazonaws/blox/scheduling/state/PaginatedResponseIterator.java:
```diff
@@ -0,0 +1,67 @@
+ * @param <T> The response type returned by the list API call
+ */
+@RequiredArgsConstructor
+class PaginatedResponseIterator<T> implements Iterator<T> {
```
why are these classes not public?

wjbuys replied:
Think it was left over from extracting this as a private inner class on one of the listers. Should it be public? I do not have strong feelings about java class visibility beyond what's needed to make the compiler happy, so I'll defer to whatever your guideline is here.

narehayrapetyan commented on scheduling-manager/src/test/java/com/amazonaws/blox/scheduling/state/ContainerInstanceListerTest.java:
```diff
@@ -0,0 +1,112 @@
+
+  @Test
+  public void returnsContainerInstancesFromAllPages() {
+    ECSAsyncClient ecs = mock(FakeECSAsyncClient.class, Mockito.CALLS_REAL_METHODS);
```
is this one of the temporary unit/integ mix tests? can we move them to a separate package so they're easy to refactor into integ tests later?

wjbuys replied:
No, this unit test has value, and will need to be expanded to cover failure cases that are hard/impossible to reproduce against a real ECS client (e.g. partial asynchronous failure)

wjbuys replied:
Oh, and to answer the second part of your question: They are pure unit tests, since they mock out all dependencies, do I don't think they belong in a separate package.

wjbuys commented:
Just to follow up, I extracted some of this PR into #278 to fix the build, so I'll have to rebase this PR after that one is merged to `dev`.

Commit:
Use lists instead of maps for tasks/instances in state snapshot

The original idea was to index tasks/instances by ID once in the
Manager, instead of having to reindex in each scheduler. However, in
practice we want to index tasks by instance id (not task id), so it's
more flexible to use lists instead.

Commit:
Make sure SingleTaskScheduler actually only starts one task

This requires persisting something in the taskGroup, so that we know
whether the scheduler started the task, or if it's unrelated.

Commit:
Implement efficient cluster state snapshots in ECSStateClient

This uses an `ECSAsyncClient` to implement the `ECSState` interface by
concurrently calling describe for sequential pages of list responses.

The integration test is very basic, but we may be able to build on it
for further integration tests that require a test cluster.

See #261 for details.

wjbuys commented:
I've rebased this PR onto dev, should be ready to review/ship now.
